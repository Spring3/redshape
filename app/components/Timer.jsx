import React, { Component } from 'react';
import moment from 'moment';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import PlayIcon from 'mdi-react/PlayIcon';
import PauseIcon from 'mdi-react/PauseIcon';
import StopIcon from 'mdi-react/StopIcon';

import actions from '../actions';

import { GhostButton } from './Button';

const ActiveTimer = styled.div`
  max-width: 100%;
  box-sizing: border-box;
  padding: 20px;
  position: fixed;
  bottom: 0;
  width: 100%;
  background: ${props => props.theme.bg};
  display: ${props => props.isEnabled ? 'flex' : 'none'};
  align-items: center;
  box-shadow: 0px 0px 15px ${props => props.theme.bgDark};

  div.buttons {
    margin: 0 20px;
  }

  div.issueName,
  div.time {
    margin: 0 20px;
    font-size: 16px;
    font-weight: bold;
  }

  div.buttons {
    a {
      padding: 10px 0px;

      &:hover {
        background: ${props => props.theme.bgDark};
      }
    }

    a:first-child {
      margin-right: 20px;
    }
  }

  div.time {
    color: ${props => props.theme.main};
  }
`;

const StyledButton = styled(GhostButton)`
  padding: 0px;
`;

class Timer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      value: props.trackedTime || props.initialValue || 0
    };

    this.interval = props.isEnabled && !props.isPaused
     ? setInterval(this.tick, 1000)
     : undefined;
  }

  tick = () => {
    const { value } = this.state;
    this.setState({ value: value + 1000 });
  }

  stopInterval = () => {
    clearInterval(this.interval);
    this.interval = undefined;
  }

  cleanup = () => {
    const { isEnabled, isPaused } = this.props;
    if (isEnabled && !isPaused) {
      this.onPause();
    }
    this.stopInterval();
    window.removeEventListener('beforeunload', this.cleanup);
  }

  componentWillMount() {
    window.addEventListener('beforeunload', this.cleanup);
  }

  componentWillUnmount() {
    this.cleanup();
  }

  componentDidUpdate(oldProps) {
    const { isEnabled } = this.props;
    if (isEnabled !== oldProps.isEnabled) {
      this.stopInterval();
      // if was disabled, but now is enabled
      if (isEnabled) {
        this.interval = setInterval(this.tick, 1000);
      }
      // otherwise, if was enabled, but now it's disabled, we don't do anything
      // because the interval was already cleared above
    }
  }

  onPause = () => {
    this.stopInterval();
    const { onPause, trackedIssue, pauseTimer } = this.props;
    const { value } = this.state;
    pauseTimer(value);
    if (onPause) {
      onPause(value, trackedIssue)
    }
  }

  onContinue = () => { 
    this.interval = setInterval(this.tick, 1000);
    const { onContinue, trackedIssue, continueTimer } = this.props;
    continueTimer();
    if (onContinue) {
      onContinue(trackedIssue);
    }
  }

  onStop = () => {
    this.stopInterval();
    const { value } = this.state;
    const { onStop, trackedIssue, stopTimer } = this.props;
    stopTimer(value);
    if (onStop) {
      onStop(value, trackedIssue);
    }
    this.setState({ value: 0 });
  }

  render() {
    const { value } = this.state;
    const { isEnabled, trackedIssue, isPaused } = this.props;
    const timeString = moment.utc(value).format('HH:mm:ss');
    return (
      <ActiveTimer isEnabled={isEnabled}>
        <div className="buttons">
          <StyledButton
            onClick={this.onStop}
          >
            <StopIcon size={35} />
          </StyledButton>
          { isPaused && (
              <StyledButton
                onClick={this.onContinue}
              >
                <PlayIcon size={35} />
              </StyledButton>
            )
          }
          { !isPaused && (
              <StyledButton
                onClick={this.onPause}
              >
                <PauseIcon size={35} />
              </StyledButton>
            )
          }
        </div>
        <div className="issueName">
          <span>{trackedIssue.subject}</span>
        </div>
        <div className="time">
          <span>{timeString}</span>
        </div>
      </ActiveTimer>
    );
  }
}

Timer.propTypes = {
  isEnabled: PropTypes.bool,
  isPaused: PropTypes.bool,
  initialValue: PropTypes.number,
  trackedIssue: PropTypes.shape({
    id: PropTypes.number.isRequired,
    subject: PropTypes.string.isRequired,
    author: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired
    }).isRequired,
    project: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired
    }).isRequired,
    activity: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired
    }).isRequired
  }).isRequired,
  trackedTime: PropTypes.number,
  onStop: PropTypes.func,
  onPause: PropTypes.func,
  onContinue: PropTypes.func,
  pauseTimer: PropTypes.func.isRequired,
  continueTimer: PropTypes.func.isRequired,
  stopTimer: PropTypes.func.isRequired
};

Timer.defaultProps = {
  initialValue: 0,
  isEnabled: false,
  isPaused: false,
  issueTitle: ''
};

const mapStateToProps = state => ({
  isEnabled: state.tracking.isEnabled,
  isPaused: state.tracking.isPaused,
  trackedTime: state.tracking.duration,
  trackedIssue: state.tracking.issue,
});

const mapDispatchToProps = dispatch => ({
  pauseTimer: value => dispatch(actions.tracking.trackingPause(value)),
  continueTimer: () => dispatch(actions.tracking.trackingContinue()),
  stopTimer: value => dispatch(actions.tracking.trackingStop(value))
});

export default connect(mapStateToProps, mapDispatchToProps)(Timer);
