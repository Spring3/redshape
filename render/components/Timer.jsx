import React, { Fragment, Component } from 'react';
import moment from 'moment';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import PlayIcon from 'mdi-react/PlayIcon';
import PauseIcon from 'mdi-react/PauseIcon';
import StopIcon from 'mdi-react/StopIcon';

import actions from '../actions';

import { GhostButton } from './Button';
import { animationSlideUp } from '../animations';
import Link from './Link';

import { setupTimerIPC, sendTimerIPC } from '../ipc';

const ActiveTimer = styled.div`
  animation: ${animationSlideUp} .7s ease-in;
  max-width: 100%;
  box-sizing: border-box;
  padding: 20px;
  position: fixed;
  bottom: 0;
  width: 100%;
  background: ${props => props.theme.bg};
  display: ${props => props.isEnabled ? 'flex' : 'none'};
  align-items: center;
  box-shadow: 0px -2px 20px ${props => props.theme.bgDark};
  border-top: 2px solid ${props => props.theme.bgDark};

  div.buttons {
    margin: 0 20px;
  }

  div.time {
    margin: 0 20px;
    font-size: 16px;
    font-weight: bold;
  }

  div.buttons {
    a {
      padding: 10px 0px;

      &:hover {
        background: ${props => props.theme.bgDark};
      }
    }

    a:first-child {
      margin-right: 20px;
    }
  }

  div.time {
    color: ${props => props.theme.main};
  }
`;

const StyledButton = styled(GhostButton)`
  padding: 0px;
`;

const MaskedLink = styled(Link)`
  color: inherit;
  padding: 0;
  margin: 0 20px;
  font-size: 16px;
  font-weight: bold;
  text-decoration: none;
`;

class Timer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      value: props.trackedTime || props.initialValue || 0,
      timestamp: null, // used only when store/restore to/from timestamp (window hidden)
    };

    setupTimerIPC(this);

    const { isEnabled, isPaused, trackedIssue } = this.props;
    if (isEnabled){
      sendTimerIPC('timer-info', {isEnabled, isPaused, issue: trackedIssue})
    }

    this.interval = props.isEnabled && !props.isPaused
     ? setInterval(this.tick, 1000)
     : undefined;
  }
  /* stop time interval and store tracked time + current datetime */
  storeToTimestamp(){
    const { timestamp } = this.state;
    const { isEnabled, isPaused, trackedIssue } = this.props;
    if (isEnabled && !isPaused){
      this.setState({
        timestamp: {
          value: this.state.value,
          datetime: moment()
        }
      });
      this.stopInterval()
    }
    sendTimerIPC('timer-info', {isEnabled, isPaused, issue: trackedIssue})
  }
  /* restore time interval based on stored tracked time + diff datetimes */
  restoreFromTimestamp(enableInterval = true){
    const { timestamp } = this.state;
    const { isEnabled, isPaused } = this.props;
    if (timestamp && isEnabled && !isPaused){
      const { value, datetime } = timestamp;
      const newValue = moment().diff(datetime, 'ms') + value;
      this.setState({ timestamp: null, value: newValue });
      if (enableInterval) {
        this.interval = setInterval(this.tick, 1000);
      }
    }
  }

  tick = () => {
    const { value } = this.state;
    this.setState({ value: value + 1000 });
  }

  stopInterval = () => {
    if (this.interval) {
      clearInterval(this.interval);
      this.interval = undefined;
    }
  }

  cleanup = () => {
    const { isEnabled, isPaused } = this.props;
    if (isEnabled && !isPaused) {
      this.onPause();
    }
    this.stopInterval();
    window.removeEventListener('beforeunload', this.cleanup);
  }

  saveState = () => {
    const { isEnabled, isPaused } = this.props;
    if (isEnabled && !isPaused) {
      const { continueTimer } = this.props;
      continueTimer(this.state.value);
    }
    this.stopInterval();
  }

  componentWillMount() {
    window.addEventListener('beforeunload', this.cleanup);
  }

  componentWillUnmount() {
    this.saveState();
  }

  componentDidUpdate(oldProps) {
    const { isEnabled, isPaused, trackedIssue } = this.props;
    if (isEnabled !== oldProps.isEnabled) {
      this.stopInterval();
      // if was disabled, but now is enabled
      if (isEnabled) {
        this.interval = setInterval(this.tick, 1000);
        sendTimerIPC('timer-info', {isEnabled, isPaused, issue: trackedIssue });
      }
      // otherwise, if was enabled, but now it's disabled, we don't do anything
      // because the interval was already cleared above
    }
  }

  onPause = () => {
    this.stopInterval();
    const { onPause, trackedIssue, pauseTimer } = this.props;
    const { value } = this.state;
    pauseTimer(value);
    if (onPause) {
      onPause(value, trackedIssue)
    }
    sendTimerIPC('timer-info', {isEnabled: true, isPaused: true, issue: trackedIssue})
  }

  onContinue = () => {
    this.interval = setInterval(this.tick, 1000);
    const { onContinue, trackedIssue, continueTimer } = this.props;
    const { value } = this.state;
    continueTimer(value);
    if (onContinue) {
      onContinue(trackedIssue, value);
    }
    sendTimerIPC('timer-info', {isEnabled: true, isPaused: false, issue: trackedIssue})
  }

  onStop = () => {
    this.stopInterval();
    const { value } = this.state;
    const { onStop, trackedIssue, stopTimer } = this.props;
    stopTimer(value);
    if (onStop) {
      onStop(value, trackedIssue);
    }
    this.setState({ value: 0 });
    sendTimerIPC('timer-info', {isEnabled: false, issue: trackedIssue})
  }

  redirectToTrackedLink = (event) => {
    event.preventDefault();
    this.props.history.push(`/app/issue/${this.props.trackedIssue.id}`);
  }

  componentWillReceiveProps(newProps) {
    const { trackedTime } = newProps;
    this.setState({ value: trackedTime })
  }

  render() {
    const { value } = this.state;
    const { isEnabled, trackedIssue, isPaused } = this.props;
    const timeString = moment.utc(value).format('HH:mm:ss');
    return (
      <Fragment>
      <ActiveTimer isEnabled={isEnabled}>
        <div className="buttons">
          <StyledButton
            onClick={this.onStop}
          >
            <StopIcon size={35} />
          </StyledButton>
          { isPaused && (
              <StyledButton
                onClick={this.onContinue}
              >
                <PlayIcon size={35} />
              </StyledButton>
            )
          }
          { !isPaused && (
              <StyledButton
                onClick={this.onPause}
              >
                <PauseIcon size={35} />
              </StyledButton>
            )
          }
        </div>
        <div className="issueName">
          { (isEnabled ?
          <MaskedLink
            href='#'
            onClick={this.redirectToTrackedLink}
          >
            {trackedIssue.subject}
          </MaskedLink>
            : null)}
        </div>
        <div className="time">
          <span>{timeString}</span>
        </div>
      </ActiveTimer>
      </Fragment>
    );
  }
}

Timer.propTypes = {
  isEnabled: PropTypes.bool,
  isPaused: PropTypes.bool,
  initialValue: PropTypes.number,
  trackedIssue: PropTypes.shape({
    id: PropTypes.number.isRequired,
    subject: PropTypes.string.isRequired,
    author: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired
    }).isRequired,
    project: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired
    }).isRequired,
    activity: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired
    }).isRequired
  }).isRequired,
  trackedTime: PropTypes.number,
  onStop: PropTypes.func,
  onPause: PropTypes.func,
  onContinue: PropTypes.func,
  pauseTimer: PropTypes.func.isRequired,
  continueTimer: PropTypes.func.isRequired,
  stopTimer: PropTypes.func.isRequired,
  history: PropTypes.object.isRequired
};

Timer.defaultProps = {
  initialValue: 0,
  isEnabled: false,
  isPaused: false,
  issueTitle: ''
};

const mapStateToProps = state => ({
  isEnabled: state.tracking.isEnabled,
  isPaused: state.tracking.isPaused,
  trackedTime: state.tracking.duration,
  trackedIssue: state.tracking.issue,
});

const mapDispatchToProps = dispatch => ({
  pauseTimer: value => dispatch(actions.tracking.trackingPause(value)),
  continueTimer: value => dispatch(actions.tracking.trackingContinue(value)),
  stopTimer: value => dispatch(actions.tracking.trackingStop(value))
});

export default connect(mapStateToProps, mapDispatchToProps)(Timer);
