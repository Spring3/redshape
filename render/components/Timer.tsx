import React, {
  useState, useEffect, useRef, useCallback
} from 'react';
// eslint-disable-next-line
import moment from 'moment';
import styled from 'styled-components';
import PlayIcon from 'mdi-react/PlayIcon';
import PauseIcon from 'mdi-react/PauseIcon';
import StopIcon from 'mdi-react/StopIcon';

import { GhostButton } from './GhostButton';
import { animationSlideUp } from '../animations';
import { Link } from './Link';
import { useOvermindActions, useOvermindState } from '../store';
import { Issue, TimeTrackingAction } from '../../types';

const ActiveTimer = styled.div`
  animation: ${animationSlideUp} .7s ease-in;
  max-width: 100%;
  box-sizing: border-box;
  padding: 20px;
  position: fixed;
  bottom: 0;
  width: 100%;
  background: ${(props) => props.theme.bg};
  display: ${(props) => (props.isEnabled ? 'flex' : 'none')};
  align-items: center;
  box-shadow: 0px -2px 20px ${(props) => props.theme.bgDark};
  border-top: 2px solid ${(props) => props.theme.bgDark};
  
  div.panel {
    flex-grow: 0;
    min-width: 520px;
    display: flex;
    align-items: center;
    max-width: ${(props) => (props.advancedTimerControls ? '900px' : '1800px')};
  }
  
  div.buttons {
    margin: 0 20px;
    display: flex;
  }
  
  div.time {
    margin: 0 20px;
    font-size: 16px;
    font-weight: bold;
  }

  div.buttons {
    a {
      padding: 10px 0px;

      &:hover {
        background: ${(props) => props.theme.bgDark};
      }
    }

    a:first-child {
      margin-right: 20px;
    }
  }
  
  div.buttons.buttons-advanced {
    a {
      margin-right: 5px;
    }
    a:last-child {
      margin-right: initial;
    }
  }
  
  div.issueName {
    padding: 0 20px;
    max-width: 500px;
  }

  div.time {
    color: ${(props) => props.theme.main};
  }
  
  input[name="comment"] {
    flex-grow: 2;
    margin-left: 20px;
    width: initial;
    border: none;
    border-radius: 0;
    border-bottom: 1px ${(props) => props.theme.bgDark} solid;
    color: #A4A4A4;
    &:focus {
      border: none;
      border-radius: 0;
      border-bottom: 1px ${(props) => props.theme.main} solid;
      box-shadow: none;
    }
  }
  
`;

const StyledButton = styled(GhostButton)`
  padding: 0px;
`;

const MaskedLink = styled(Link)`
  color: inherit;
  padding: 0;
  font-size: 16px;
  font-weight: bold;
  text-decoration: none;
`;

type TimerEventHook = (args: { issue: Issue, recordedTime: number }) => void;

type TimerProps = {
  startFrom?: number;
  autoStart?: boolean;
  onStop?: TimerEventHook;
  onStart?: TimerEventHook;
  onPause?: TimerEventHook;
  onContinue?: TimerEventHook;
}

const Timer = ({ startFrom = 0, autoStart = false, onStop, onStart, onPause, onContinue }: TimerProps) => {
  const [timeMs, setTimeMs] = useState(startFrom);
  const interval = useRef<ReturnType<typeof setInterval>>();
  const state = useOvermindState();
  const actions = useOvermindActions();

  const [lastTimeRecord] = state.timeTracking.records.slice(-1);
  const issueId = lastTimeRecord?.issueId;
  const targetIssue = state.issues.byId[issueId];

  const isCounting = [TimeTrackingAction.CONTINUE, TimeTrackingAction.START].includes(lastTimeRecord?.action);
  const isPaused = lastTimeRecord?.action === TimeTrackingAction.PAUSE;
  const isStopped = lastTimeRecord?.action === TimeTrackingAction.STOP;

  const tick = useCallback(() => {
    setTimeMs((trackedTimeMs) => trackedTimeMs + 1000);
  }, []);

  const start = useCallback(() => {
    if (isPaused || isStopped) {
      actions.timeTracking.track({ issueId });
      if (!interval.current) {
        interval.current = setInterval(tick, 1000);
      }
      if (onStart) {
        onStart({ issue: targetIssue, recordedTime: timeMs });
      }
    }
  }, [actions.timeTracking.track, issueId, tick, targetIssue, onStart, timeMs]);

  const stop = useCallback(() => {
    if (isCounting) {
      actions.timeTracking.stop();
      if (interval.current) {
        clearInterval(interval.current);
      }
      if (onStop) {
        onStop({ issue: targetIssue, recordedTime: timeMs });
      }
    }
  }, [actions.timeTracking.stop, onStop, targetIssue, timeMs]);

  const pause = useCallback(() => {
    if (isCounting) {
      actions.timeTracking.pause();
      if (interval.current) {
        clearInterval(interval.current);
      }
      if (onPause) {
        onPause({ issue: targetIssue, recordedTime: timeMs });
      }
    }
  }, [actions.timeTracking.pause, onPause, targetIssue, timeMs]);

  const unpaunse = useCallback(() => {
    if (isPaused) {
      actions.timeTracking.unpause();
      if (!interval.current) {
        interval.current = setInterval(tick, 1000);
      }
      if (onContinue) {
        onContinue({ issue: targetIssue, recordedTime: timeMs });
      }
    }
  }, [actions.timeTracking.unpause, tick, onContinue, targetIssue, timeMs]);

  useEffect(() => {
    if (isCounting) {
      setTimeMs(startFrom);
      if (autoStart) {
        start();
      }
    }
  }, [startFrom, autoStart, start]);

  const timeString = moment.utc(timeMs).format('HH:mm:ss');
  return (
    <>
      <ActiveTimer isEnabled={lastTimeRecord && !isStopped}>
        <div className="panel">
          <div className="buttons">
            <StyledButton
              id="stop-timer"
              onClick={stop}
            >
              <StopIcon size={35} />
            </StyledButton>
            { isPaused
              ? (
                <StyledButton
                  id="continue-timer"
                  onClick={unpaunse}
                >
                  <PlayIcon size={35} />
                </StyledButton>
              )
              : (
                <StyledButton
                  id="pause-timer"
                  onClick={pause}
                >
                  <PauseIcon size={35} />
                </StyledButton>
              )}
          </div>
          <div className="issueName">
            { !isStopped
              ? (
                <MaskedLink
                  href="#"
                >
                  {targetIssue?.subject}
                </MaskedLink>
              )
              : null}
          </div>
          <div className="time">
            <span>{timeString}</span>
          </div>
        </div>
      </ActiveTimer>
    </>
  );
};

export {
  Timer
};
