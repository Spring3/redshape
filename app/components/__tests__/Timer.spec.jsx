import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

// using mount instead of shallow due to https://github.com/airbnb/enzyme/issues/1908

import Timer from '../Timer';

const mockStore = configureStore([thunk]);

const waitSeconds = (n = 1) => new Promise(resolve => setTimeout(() => resolve(), n * 1000));

describe('Timer component', () => {
  it('should match the snapshot', () => {
    const state = {
      tracking: {
        isEnabled: true,
        isPaused: true,
        duration: 0,
        issue: {
          id: '123abc',
          subject: 'Test issue'
        }
      }
    };
    const store = mockStore(state);
    const tree = renderer.create(
      <Provider store={store}>
        <Timer
          trackedDuration={4000}
        />
      </Provider>,
      { context: { store } }
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should automatically resume timer if trackedTime prop was provided', async () => {
    const onStop = jest.fn();
    const onPause = jest.fn();
    const onContinue = jest.fn();

    const state = {
      tracking: {
        isEnabled: true,
        isPaused: false,
        duration: 4000,
        issue: {
          id: '123abc',
          subject: 'Test issue'
        }
      }
    };

    const store = mockStore(state);

    const wrapper = mount(
      <Provider store={store}>
        <Timer
          onStop={onStop}
          onPause={onPause}
          onContinue={onContinue}
        />
      </Provider>,
      { context: { store } }
    );

    const timer = wrapper.find('Timer').instance();
    expect(timer).toBeTruthy();
    expect(timer.interval).toBeDefined();
    await waitSeconds(1);
    expect(timer.state.value).toBe(5000);
    expect(wrapper.find('.buttons').children().length).toBe(2);
    expect(wrapper.find('.buttons').childAt(1).exists('PauseIcon')).toBe(true);
    wrapper.find('.buttons').childAt(1).find('GhostButton').simulate('click');
    wrapper.update();
    expect(timer.interval).not.toBeDefined();
  });

  it('should start the counter from the initialValue if given', async () => {
    const onStop = jest.fn();
    const onPause = jest.fn();
    const onContinue = jest.fn();

    const state = {
      tracking: {
        isEnabled: true,
        isPaused: false,
        duration: 0,
        issue: {
          id: '123abc',
          subject: 'Test issue'
        }
      }
    };

    const store = mockStore(state);

    const wrapper = mount(
      <Provider store={store}>
        <Timer
          initialValue={3000}
          onStop={onStop}
          onPause={onPause}
          onContinue={onContinue}
        />
      </Provider>,
      { context: { store } }
    );

    const timer = wrapper.find('Timer').instance();
    expect(timer).toBeTruthy();

    expect(timer.interval).toBeDefined();
    expect(timer.state.value).toBe(3000);
    await waitSeconds(1);
    expect(timer.state.value).toBe(4000);
    wrapper.unmount();
  });

  it('should automatically save the progress before unmount', async () => {
    const onStop = jest.fn();
    const onPause = jest.fn();
    const onContinue = jest.fn();

    const state = {
      tracking: {
        isEnabled: true,
        isPaused: false,
        duration: 4000,
        issue: {
          id: '123abc',
          subject: 'Test issue'
        }
      }
    };

    const store = mockStore(state);

    const wrapper = mount(
      <Provider store={store}>
        <Timer
          onStop={onStop}
          onPause={onPause}
          onContinue={onContinue}
        />
      </Provider>,
      { context: { store } }
    );

    const timer = wrapper.find('Timer').instance();
    expect(timer).toBeTruthy();

    const cleanupSpy = jest.spyOn(timer, 'cleanup');

    expect(timer.interval).toBeDefined();
    await waitSeconds(1);
    expect(timer.state.value).toBe(5000);
    wrapper.unmount();

    expect(cleanupSpy).toHaveBeenCalled();
    expect(onPause).toHaveBeenCalledWith(5000, state.tracking.issue);
  });

  it('should allow to pause the timer', async () => {
    const onStop = jest.fn();
    const onPause = jest.fn();
    const onContinue = jest.fn();

    const state = {
      tracking: {
        isEnabled: true,
        isPaused: false,
        duration: 0,
        issue: {
          id: '123abc',
          subject: 'Test issue'
        }
      }
    };

    const store = mockStore(state);

    const wrapper = mount(
      <Provider store={store}>
        <Timer
          onStop={onStop}
          onPause={onPause}
          onContinue={onContinue}
        />
      </Provider>,
      { context: { store } }
    );

    const timer = wrapper.find('Timer').instance();
    expect(timer).toBeTruthy();
    expect(timer.interval).toBeDefined();

    await waitSeconds(1);
    expect(wrapper.exists('.buttons')).toBe(true);
    expect(wrapper.find('.buttons').children().length).toBe(2);
    expect(wrapper.find('.buttons').childAt(1).exists('PauseIcon')).toBe(true);

    wrapper.find('.buttons').childAt(1).find('GhostButton').simulate('click');
    wrapper.update();

    expect(onPause).toHaveBeenCalledWith(1000, state.tracking.issue);
  });

  it('should allow to resume the paused timer', async () => {
    const onStop = jest.fn();
    const onPause = jest.fn();
    const onContinue = jest.fn();

    const state = {
      tracking: {
        isEnabled: true,
        isPaused: true,
        duration: 0,
        issue: {
          id: '123abc',
          subject: 'Test issue'
        }
      }
    };

    const store = mockStore(state);

    const wrapper = mount(
      <Provider store={store}>
        <Timer
          onStop={onStop}
          onPause={onPause}
          onContinue={onContinue}
        />
      </Provider>,
      { context: { store } }
    );

    const timer = wrapper.find('Timer').instance();
    expect(timer).toBeTruthy();

    const timerTick = jest.spyOn(timer, 'tick');
    expect(timer.interval).not.toBeDefined();

    expect(wrapper.exists('.buttons')).toBe(true);
    expect(wrapper.find('.buttons').children().length).toBe(2);
    expect(wrapper.find('.buttons').childAt(1).exists('PlayIcon')).toBe(true);

    wrapper.find('.buttons').childAt(1).find('GhostButton').simulate('click');
    await waitSeconds(1);
    wrapper.update();
    expect(onContinue).toHaveBeenCalledTimes(1);
    expect(timerTick).toHaveBeenCalledTimes(1);
    expect(timer.state.value).toBe(1000);
    expect(timer.interval).toBeDefined();
    wrapper.unmount();
  });

  it('should be able to stop time timer', async () => {
    const onStop = jest.fn();
    const onPause = jest.fn();
    const onContinue = jest.fn();

    const state = {
      tracking: {
        isEnabled: true,
        isPaused: false,
        duration: 0,
        issue: {
          id: '123abc',
          subject: 'Test issue'
        }
      }
    };

    const store = mockStore(state);

    const wrapper = mount(
      <Provider store={store}>
        <Timer
          onStop={onStop}
          onPause={onPause}
          onContinue={onContinue}
        />
      </Provider>,
      { context: { store } }
    );

    const timer = wrapper.find('Timer').instance();
    expect(timer).toBeTruthy();

    expect(timer.interval).toBeDefined();
    expect(wrapper.exists('.buttons')).toBe(true);
    expect(wrapper.find('.buttons').children().length).toBe(2);
    expect(wrapper.find('.buttons').childAt(0).exists('StopIcon')).toBe(true);

    await waitSeconds(1);
    wrapper.find('.buttons').childAt(0).find('GhostButton').simulate('click');
    wrapper.update();

    expect(onStop).toHaveBeenCalledWith(1000, state.tracking.issue);
    expect(timer.interval).not.toBeDefined();

    expect(onStop).toHaveBeenCalledTimes(1);
  });
});
