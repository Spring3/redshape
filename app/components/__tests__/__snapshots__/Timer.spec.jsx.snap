// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer component should match the snapshot 1`] = `
.c2 {
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: color ease;
  transition: color ease;
}

.c2:active,
.c2:focus,
.c2:visited {
  background: transparent;
}

.c0 {
  max-width: 100%;
  box-sizing: border-box;
  padding: 20px;
  position: fixed;
  bottom: 0;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-shadow: 0px 0px 15px;
}

.c0 div.buttons {
  margin: 0 20px;
}

.c0 div.issueName,
.c0 div.time {
  margin: 0 20px;
  font-size: 16px;
  font-weight: bold;
}

.c0 div.buttons a {
  padding: 10px 0px;
}

.c0 div.buttons a:first-child {
  margin-right: 20px;
}

.c1 {
  padding: 0px;
}

<div
  className="c0"
>
  <div
    className="buttons"
  >
    <a
      className="c1 c2"
      disabled={false}
      href="#"
      onClick={[Function]}
    >
      <svg
        className="mdi-icon "
        fill="currentColor"
        height={35}
        viewBox="0 0 24 24"
        width={35}
      >
        <path
          d="M18,18H6V6H18V18Z"
        />
      </svg>
    </a>
    <a
      className="c1 c2"
      disabled={false}
      href="#"
      onClick={[Function]}
    >
      <svg
        className="mdi-icon "
        fill="currentColor"
        height={35}
        viewBox="0 0 24 24"
        width={35}
      >
        <path
          d="M8,5.14V19.14L19,12.14L8,5.14Z"
        />
      </svg>
    </a>
  </div>
  <div
    className="issueName"
  >
    <span>
      Test issue
    </span>
  </div>
  <div
    className="time"
  >
    <span>
      00:00:00
    </span>
  </div>
</div>
`;
